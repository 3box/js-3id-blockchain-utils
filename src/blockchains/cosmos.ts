import { BlockchainHandler, BlockchainHandlerOpts } from '../blockchain-handler';
import { AccountID } from 'caip';
import { getConsentMessage, LinkProof } from '../utils';
import { signTx, Tx, SignMeta, verifyTx, createAddress } from '@tendermint/sig';
const { base64ToBytes, bytesToBase64, toCanonicalJSONBytes } = require('@tendermint/belt');
import * as uint8arrays from 'uint8arrays';

const namespace = 'cosmos';

const stringHex = (str: string): string => `0x${uint8arrays.toString(uint8arrays.fromString(str), 'base16')}`;

// return data in the cosmos unsigned transaction format
function asTransaction(address: string, message: string): Tx {
  return {
    fee: {
      amount: [{ amount: '0', denom: '' }],
      gas: '0',
    },
    memo: message,
    msg: [
      {
        type: 'cosmos-sdk/MsgSend',
        value: {
          from_address: address,
          to_address: address,
          amount: [{ amount: '0', denom: '0' }],
        },
      },
    ],
  };
}

// generate metadata for signing the transaction
function getMetaData(): SignMeta {
  return {
    account_number: '1',
    chain_id: 'cosmos',
    sequence: '0',
  };
}

async function createLink(did: string, account: AccountID, provider: any, opts: BlockchainHandlerOpts): Promise<LinkProof> {
  const { message, timestamp } = getConsentMessage(did, !opts?.skipTimestamp);
  const linkMessageHex = stringHex(message);
  const res = await signTx(asTransaction(account.address, linkMessageHex), getMetaData(), provider);
  const signature = bytesToBase64(toCanonicalJSONBytes(res));
  const proof: LinkProof = {
    version: 2,
    type: 'eoa',
    message: linkMessageHex,
    signature,
    account: account.toString(),
  };
  if (!opts?.skipTimestamp) proof.timestamp = timestamp;
  return proof;
}

async function authenticate(message: string, account: AccountID, provider: any): Promise<string> {
  const linkMessageHex = stringHex(message);
  const res = await signTx(asTransaction(account.address, linkMessageHex), getMetaData(), provider);
  return bytesToBase64(toCanonicalJSONBytes(res));
}

async function validateLink(proof: LinkProof): Promise<LinkProof | null> {
  const payload = JSON.parse(Buffer.from(base64ToBytes(proof.signature)).toString());
  const message = stringHex(proof.message);
  const signer_address = createAddress(base64ToBytes(payload.signatures[0].pub_key.value));
  const account = new AccountID(proof.account);
  const is_sig_valid = verifyTx(payload, getMetaData());
  if (is_sig_valid && payload.memo === message && signer_address === account.address) {
    return proof;
  } else {
    return null;
  }
}

const Handler: BlockchainHandler = {
  namespace,
  authenticate,
  validateLink,
  createLink,
};

export default Handler;
