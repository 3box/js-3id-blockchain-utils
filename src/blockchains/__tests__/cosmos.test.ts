import { AccountID } from "caip";
import { createWalletFromMnemonic } from '@tendermint/sig';
import {authenticate, createLink, validateLink} from "../cosmos";

const chainId = "cosmoshub-3";

const did = 'did:3:bafysdfwefwe'
const mnemonic = 'test salon husband push melody usage fine ensure blade deal miss twin';
const localProvider = createWalletFromMnemonic(mnemonic);

const addressToAccountID = (address: string): AccountID => new AccountID(`${address}@${chainId}`)

describe('Blockchain: Cosmos', () => {

    describe('createLink', () => {
        expect(true)
    })

})
