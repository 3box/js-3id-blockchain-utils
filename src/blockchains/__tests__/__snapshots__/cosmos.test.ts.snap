// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain: Cosmos createLink create proof for cosmoshub3 1`] = `
Object {
  "account": "cosmos1gs6gu0w9xvqcfhm3nlhnmqkf8aphnsxyjlgpqd@cosmos:cosmoshub-3",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe",
  "signature": "eyJwdWJfa2V5Ijp7InR5cGUiOiJ0ZW5kZXJtaW50L1B1YktleVNlY3AyNTZrMSIsInZhbHVlIjoiQTQxWEVNTHROeG9uaGVTYXA1ODk3N2RzRmFJOHlVa3c0d1ZLK2JHd080SXIifSwic2lnbmF0dXJlIjoicUw3d2NISkNKeXNOajF2VzNlUlRGb0xMd3oyZENvdVZINGorUkpsVnoyb2JmVFNuR0ZEcER6MlpRajJlanphYlpDNGl0VTVQYjQxOTdwSHF4V3hSNEE9PSJ9",
  "type": "eoa",
  "version": 2,
}
`;
